src/box/lua/upgrade.lua

```diff
@@ -1517,6 +1530,405 @@ local function upgrade()
+-- Downgrade part
+--------------------------------------------------------------------------------
+
+--------------------------------------------------------------------------------
+-- Tarantool 2.9.1
+--------------------------------------------------------------------------------
+
+local function restore_sql_builtin_functions(issue_handler)
+    if issue_handler.dry_run then
+        return
+    end
+    local sql_builtin_list = {
+        "TRIM", "TYPEOF", "PRINTF", "UNICODE", "CHAR", "HEX", "VERSION",
+        "QUOTE", "REPLACE", "SUBSTR", "GROUP_CONCAT", "JULIANDAY", "DATE",
+        "TIME", "DATETIME", "STRFTIME", "CURRENT_TIME", "CURRENT_TIMESTAMP",
+        "CURRENT_DATE", "LENGTH", "POSITION", "ROUND", "UPPER", "LOWER",
+        "IFNULL", "RANDOM", "CEIL", "CEILING", "CHARACTER_LENGTH",
+        "CHAR_LENGTH", "FLOOR", "MOD", "OCTET_LENGTH", "ROW_COUNT", "COUNT",
+        "LIKE", "ABS", "EXP", "LN", "POWER", "SQRT", "SUM", "TOTAL", "AVG",
+        "RANDOMBLOB", "NULLIF", "ZEROBLOB", "MIN", "MAX", "COALESCE", "EVERY",
+        "EXISTS", "EXTRACT", "SOME", "GREATER", "LESSER", "SOUNDEX",
+        "LIKELIHOOD", "LIKELY", "UNLIKELY", "_sql_stat_get", "_sql_stat_push",
+        "_sql_stat_init", "GREATEST", "LEAST",
+    }
+    local datetime = os.date("%Y-%m-%d %H:%M:%S")
+    local _func = box.space._func
+    -- Otherwise we can't insert SQL_BUILTIN function. It is prohibited
+    -- to add since 2.9.0.
+    set_system_triggers(false)
+    local status, err = pcall(function()
+        for _, func in ipairs(sql_builtin_list) do
+            if _func.index.name:get(func) == nil then
+                local t = _func:auto_increment{
+                    ADMIN, func, 1, 'SQL_BUILTIN', '', 'function', {}, 'any',
+                    'none', 'none', false, false, true, {}, setmap({}), '',
+                    datetime, datetime}
+                box.space._priv:replace{ADMIN, PUBLIC, 'function', t.id,
+                                        box.priv.X}
+            end
+        end
+    end)
+    if not status then
+        pcall(set_system_triggers, true)
+        error(err)
+    end
+    set_system_triggers(true)
+end
+
+local function downgrade_from_2_9_1(issue_handler)
+    restore_sql_builtin_functions(issue_handler)
+end
+
+--------------------------------------------------------------------------------
+-- Tarantool 2.10.0
+--------------------------------------------------------------------------------
+
+local function remove_deferred_deletes(issue_handler)
+    if issue_handler.dry_run then
+        return
+    end
+    log.info("remove defer_deletes from space options")
+    for _, row in box.space._space:pairs() do
+        local new_flags = table.copy(row.flags)
+        if new_flags.defer_deletes ~= nil then
+            new_flags.defer_deletes = nil
+            setmap(new_flags)
+            box.space._space:update(row.id, {{'=', 'flags', new_flags}})
```

Comment:
Nit: please rename `row` to `space` in all `box.space._space:pairs` calls.

What if one of the updates fails — isn't the database left in an inconsistent space? Shouldn't we use a transaction here? Same question for DML operations of other handlers.

Reply:
AFAIK transactions do not support  DDL thus we don't have transactions nor in upgrade neither in downgrade. At least you have idempotency of downgrade. So when problem is fixed you can run downgrade again and finish it.

Reply:
@alyapunov RFC